File: sdk/python/dynaep/temporal/async_bridge_clock.py
# ===========================================================================
# dynaep.temporal.async_bridge_clock - Async Bridge Clock with Slewing
# OPT-008: NTP sync runs in a background thread (never blocks event
# processing). Clock corrections applied via slewing (gradual adjustment)
# instead of stepping for corrections between 1ms and 1000ms.
# ===========================================================================

from __future__ import annotations
import time
import socket
import struct
import os
import threading
import logging
from dataclasses import dataclass
from typing import Optional, List, Callable

logger = logging.getLogger("dynaep.temporal.async_bridge_clock")

NTP_UNIX_EPOCH_DELTA_SECONDS = 2208988800
NTP_PACKET_SIZE = 48
NTP_REQUEST_HEADER = 0x1B
DEFAULT_NTP_SERVER = "pool.ntp.org"
NTP_PORT = 123
PTP_OFFSET_PATH = "/sys/class/ptp/ptp0/offset"
NTP_TIMEOUT_SECONDS = 5.0
SLEW_IGNORE_THRESHOLD_MS = 1.0
SLEW_STEP_THRESHOLD_MS = 1000.0


@dataclass
class ClockConfig:
    protocol: str  # "ntp" | "ptp" | "system"
    source: str
    sync_interval_ms: int
    max_drift_ms: int
    bridge_is_authority: bool


@dataclass
class SyncResult:
    success: bool
    offset_ms: float
    latency_ms: float


class AsyncBridgeClock:
    """Bridge clock with async NTP sync and slewing corrections.

    Thread safety: now() is thread-safe (reads only from atomic-like state).
    Sync runs in a daemon thread and never blocks the main thread.
    """

    def __init__(self, config: ClockConfig) -> None:
        self._config = config
        self._active_protocol = config.protocol
        self._sync_in_progress = False
        self._sync_count = 0
        self._last_sync_success = False
        self._last_sync_at = 0.0
        self._started_at = time.monotonic()

        # Offsets
        self._epoch_offset = time.time() - time.monotonic()
        self._ntp_offset_ms = 0.0

        # Slewing
        self._slew_remaining_ms = 0.0
        self._slew_start_mono = 0.0
        self._slew_duration_ms = 0.0
        self._slew_total_ms = 0.0

        # Monotonicity
        self._last_returned_ms = 0.0
        self._lock = threading.Lock()

        # Timer thread
        self._timer: Optional[threading.Timer] = None
        self._running = False

        # Event listeners
        self._listeners: List[Callable] = []

    def start(self) -> None:
        """Begin periodic sync. Initial sync is synchronous; subsequent are async."""
        self._running = True
        self._perform_sync()
        if self._config.sync_interval_ms > 0:
            self._schedule_next_sync()

    def stop(self) -> None:
        self._running = False
        if self._timer:
            self._timer.cancel()
            self._timer = None

    def now(self) -> float:
        """Return bridge-authoritative time in milliseconds.

        GUARANTEES:
        - Non-blocking (pure arithmetic)
        - Monotonically non-decreasing
        - Includes active slew adjustment
        """
        mono = time.monotonic()

        # Slew adjustment
        slew = 0.0
        if self._slew_remaining_ms != 0 and self._slew_duration_ms > 0:
            elapsed = (mono - self._slew_start_mono) * 1000  # to ms
            if elapsed >= self._slew_duration_ms:
                slew = self._slew_total_ms
                self._slew_remaining_ms = 0
            else:
                fraction = elapsed / self._slew_duration_ms
                slew = self._slew_total_ms * fraction

        bridge_ms = (mono + self._epoch_offset) * 1000 + self._ntp_offset_ms + slew

        # Monotonicity
        with self._lock:
            if bridge_ms < self._last_returned_ms:
                bridge_ms = self._last_returned_ms
            else:
                self._last_returned_ms = bridge_ms

        return bridge_ms

    def on_sync(self, listener: Callable) -> None:
        self._listeners.append(listener)

    def is_synced(self) -> bool:
        if self._last_sync_at == 0:
            return False
        elapsed = (time.time() - self._last_sync_at) * 1000
        return elapsed < self._config.sync_interval_ms * 3

    def get_active_protocol(self) -> str:
        return self._active_protocol

    def get_offset_ms(self) -> float:
        return self._ntp_offset_ms

    def measure_drift(self, agent_time_ms: float) -> float:
        return self.now() - agent_time_ms

    # -----------------------------------------------------------------------
    # Internal sync
    # -----------------------------------------------------------------------

    def _schedule_next_sync(self) -> None:
        if not self._running:
            return
        interval_sec = self._config.sync_interval_ms / 1000.0
        self._timer = threading.Timer(interval_sec, self._async_sync_wrapper)
        self._timer.daemon = True
        self._timer.start()

    def _async_sync_wrapper(self) -> None:
        """Runs in a daemon thread."""
        try:
            self._perform_sync()
        except Exception:
            logger.warning("NTP sync failed, retrying on next interval")
        finally:
            self._schedule_next_sync()

    def _perform_sync(self) -> None:
        if self._sync_in_progress:
            return
        self._sync_in_progress = True
        try:
            result = SyncResult(success=False, offset_ms=0, latency_ms=0)

            if self._active_protocol == "ptp":
                result = self._sync_via_ptp()
                if not result.success:
                    result = self._sync_via_ntp()
                    if result.success:
                        self._active_protocol = "ntp"
            elif self._active_protocol == "ntp":
                result = self._sync_via_ntp()
            else:
                result = self._sync_via_system()

            if not result.success:
                result = self._sync_via_system()
                self._active_protocol = "system"

            if result.success:
                self._apply_correction(result.offset_ms)
                self._last_sync_at = time.time()
                self._last_sync_success = True
                self._sync_count += 1
                for listener in self._listeners:
                    try:
                        listener({
                            "source": self._active_protocol,
                            "offset_ms": result.offset_ms,
                            "synced_at": self._last_sync_at,
                        })
                    except Exception:
                        pass
            else:
                self._last_sync_success = False
                logger.warning("NTP sync failed, keeping previous offset: %s", self._ntp_offset_ms)
        finally:
            self._sync_in_progress = False

    def _apply_correction(self, new_offset_ms: float) -> None:
        correction = new_offset_ms - self._ntp_offset_ms
        abs_correction = abs(correction)

        if abs_correction < SLEW_IGNORE_THRESHOLD_MS:
            return

        if abs_correction > SLEW_STEP_THRESHOLD_MS:
            self._ntp_offset_ms = new_offset_ms
            self._slew_remaining_ms = 0
            self._slew_total_ms = 0
            logger.warning("Large clock correction (step): %.1fms", correction)
            return

        # Complete any previous slew
        if self._slew_remaining_ms != 0:
            self._ntp_offset_ms += self._slew_total_ms
            self._slew_remaining_ms = 0

        self._slew_total_ms = correction
        self._slew_remaining_ms = correction
        self._slew_start_mono = time.monotonic()
        self._slew_duration_ms = self._config.sync_interval_ms

    def _sync_via_ntp(self) -> SyncResult:
        server = self._config.source or DEFAULT_NTP_SERVER
        t0 = time.time() * 1000
        try:
            sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            sock.settimeout(NTP_TIMEOUT_SECONDS)
            packet = b"\x1b" + b"\x00" * 47
            sock.sendto(packet, (server, NTP_PORT))
            data, _ = sock.recvfrom(NTP_PACKET_SIZE)
            t3 = time.time() * 1000
            sock.close()

            if len(data) < NTP_PACKET_SIZE:
                return SyncResult(False, 0, 0)

            t1 = self._extract_ntp_ts(data, 32)
            t2 = self._extract_ntp_ts(data, 40)
            offset = ((t1 - t0) + (t2 - t3)) / 2
            latency = t3 - t0
            return SyncResult(True, offset, latency)
        except Exception:
            return SyncResult(False, 0, 0)

    @staticmethod
    def _extract_ntp_ts(data: bytes, offset: int) -> float:
        seconds = struct.unpack("!I", data[offset:offset + 4])[0]
        fraction = struct.unpack("!I", data[offset + 4:offset + 8])[0]
        unix_seconds = seconds - NTP_UNIX_EPOCH_DELTA_SECONDS
        frac_ms = (fraction / 0x100000000) * 1000
        return unix_seconds * 1000 + frac_ms

    def _sync_via_ptp(self) -> SyncResult:
        try:
            with open(PTP_OFFSET_PATH, "r") as f:
                raw = f.read().strip()
            ns = int(raw)
            return SyncResult(True, ns / 1_000_000, 0)
        except Exception:
            return SyncResult(False, 0, 0)

    def _sync_via_system(self) -> SyncResult:
        logger.warning("Using system clock fallback")
        return SyncResult(True, 0, 0)

File: sdk/typescript/src/temporal/AsyncBridgeClock.ts
// ===========================================================================
// @dynaep/core - Async Bridge Clock with Slewing
// OPT-008: NTP sync is fully asynchronous (never blocks event processing).
// Clock corrections are applied via slewing (gradual adjustment) instead
// of stepping (instant jump) for corrections between 1ms and 1000ms.
//
// The now() method is synchronous, non-blocking, and monotonically
// non-decreasing. It completes in < 0.01ms (pure arithmetic).
// ===========================================================================

import * as dgram from "dgram";
import * as fs from "fs";
import type { ClockConfig, BridgeTimestamp, ClockHealth, SyncResult } from "./clock";
import type { ClockSyncEvent } from "./events";
import { createClockSyncEvent } from "./events";

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const NTP_UNIX_EPOCH_DELTA_SECONDS = 2208988800;
const NTP_PACKET_SIZE = 48;
const NTP_REQUEST_HEADER = 0x1b;
const DEFAULT_NTP_SERVER = "pool.ntp.org";
const NTP_PORT = 123;
const PTP_OFFSET_PATH = "/sys/class/ptp/ptp0/offset";
const NTP_TIMEOUT_MS = 5000;

/** Corrections smaller than this are ignored (within noise). */
const SLEW_IGNORE_THRESHOLD_MS = 1.0;

/** Corrections larger than this are stepped immediately. */
const SLEW_STEP_THRESHOLD_MS = 1000.0;

// ---------------------------------------------------------------------------
// AsyncBridgeClock
// ---------------------------------------------------------------------------

export class AsyncBridgeClock {
  private readonly config: ClockConfig;
  private activeProtocol: "ntp" | "ptp" | "system";
  private syncIntervalHandle: ReturnType<typeof setInterval> | null;
  private syncInProgress: boolean;
  private syncCount: number;
  private lastSyncSuccess: boolean;
  private lastSyncAt: number;
  private startedAt: number;

  // Time offsets
  private epochOffset: number;
  private currentNtpOffset: number;

  // Slewing state
  private slewRemainingMs: number;
  private slewStartPerf: number;
  private slewDurationMs: number;
  private slewTotalMs: number;

  // Monotonicity enforcement
  private lastReturnedMs: number;

  // Event listeners
  private readonly eventListeners: ((event: ClockSyncEvent) => void)[];

  constructor(config: ClockConfig) {
    this.config = Object.freeze({ ...config });
    this.activeProtocol = config.protocol;
    this.syncIntervalHandle = null;
    this.syncInProgress = false;
    this.syncCount = 0;
    this.lastSyncSuccess = false;
    this.lastSyncAt = 0;
    this.startedAt = performance.now();

    // Compute epoch offset: difference between performance.now() origin and Unix epoch
    // This is computed once and remains stable
    this.epochOffset = Date.now() - performance.now();
    this.currentNtpOffset = 0;

    // Slewing: no active slew
    this.slewRemainingMs = 0;
    this.slewStartPerf = 0;
    this.slewDurationMs = 0;
    this.slewTotalMs = 0;

    this.lastReturnedMs = 0;
    this.eventListeners = [];
  }

  // -------------------------------------------------------------------------
  // Lifecycle
  // -------------------------------------------------------------------------

  /**
   * Begin periodic async synchronization. Performs initial sync immediately,
   * then schedules recurring syncs at the configured interval.
   * NTP sync never blocks the event loop.
   */
  async start(): Promise<void> {
    await this.performAsyncSync();

    if (this.config.syncIntervalMs > 0) {
      this.syncIntervalHandle = setInterval(() => {
        // Fire-and-forget: never blocks
        this.performAsyncSync().catch(() => {
          // Sync failure handled internally
        });
      }, this.config.syncIntervalMs);
    }
  }

  /**
   * Stop periodic synchronization.
   */
  stop(): void {
    if (this.syncIntervalHandle !== null) {
      clearInterval(this.syncIntervalHandle);
      this.syncIntervalHandle = null;
    }
  }

  // -------------------------------------------------------------------------
  // Timestamp Production
  // -------------------------------------------------------------------------

  /**
   * Return bridge-authoritative time in milliseconds.
   *
   * GUARANTEES:
   * - Synchronous: no await, no Promise, no I/O
   * - Non-blocking: pure arithmetic only
   * - Monotonically non-decreasing: never returns a lower value
   * - Includes active slew adjustment
   * - Completes in < 0.01ms
   *
   * Formula: performance.now() + epochOffset + currentNtpOffset + activeSlewAdjustment
   */
  now(): number {
    const perfNow = performance.now();

    // Compute slew adjustment
    let slewAdjustment = 0;
    if (this.slewRemainingMs !== 0 && this.slewDurationMs > 0) {
      const elapsed = perfNow - this.slewStartPerf;
      if (elapsed >= this.slewDurationMs) {
        // Slew complete: apply full correction
        slewAdjustment = this.slewTotalMs;
        this.slewRemainingMs = 0;
      } else {
        // Proportional slew: apply fraction of total correction
        const fraction = elapsed / this.slewDurationMs;
        slewAdjustment = this.slewTotalMs * fraction;
      }
    }

    let bridgeTimeMs = perfNow + this.epochOffset + this.currentNtpOffset + slewAdjustment;

    // Enforce monotonicity: never return a value less than previous
    // If slewing would cause backward movement, hold at last value
    if (bridgeTimeMs < this.lastReturnedMs) {
      bridgeTimeMs = this.lastReturnedMs;
    } else {
      this.lastReturnedMs = bridgeTimeMs;
    }

    return bridgeTimeMs;
  }

  /**
   * Produce a BridgeTimestamp with drift calculation and monotonicity.
   */
  stamp(agentTimeMs?: number | null): BridgeTimestamp {
    const bridgeTimeMs = this.now();
    const resolvedAgentTime = agentTimeMs !== undefined ? agentTimeMs : null;
    const driftMs = resolvedAgentTime !== null ? bridgeTimeMs - resolvedAgentTime : 0;

    return {
      bridgeTimeMs,
      agentTimeMs: resolvedAgentTime,
      driftMs,
      source: this.activeProtocol,
      syncedAt: this.lastSyncAt,
    };
  }

  // -------------------------------------------------------------------------
  // Health and Status
  // -------------------------------------------------------------------------

  health(): ClockHealth {
    const perfNow = performance.now();
    const uptimeMs = perfNow - this.startedAt;

    return {
      synced: this.isSynced(),
      lastSyncAt: this.lastSyncAt,
      currentOffsetMs: this.currentNtpOffset,
      protocol: this.activeProtocol,
      source: this.config.source || DEFAULT_NTP_SERVER,
      uptimeMs,
    };
  }

  isSynced(): boolean {
    if (this.lastSyncAt === 0) return false;
    const elapsed = Date.now() - this.lastSyncAt;
    return elapsed < this.config.syncIntervalMs * 3;
  }

  getActiveProtocol(): "ntp" | "ptp" | "system" {
    return this.activeProtocol;
  }

  getOffsetMs(): number {
    return this.currentNtpOffset;
  }

  getSyncCount(): number {
    return this.syncCount;
  }

  isAuthority(): boolean {
    return this.config.bridgeIsAuthority === true;
  }

  measureDrift(agentTimeMs: number): number {
    return this.now() - agentTimeMs;
  }

  /**
   * Register a listener for AEP_CLOCK_SYNC events.
   */
  onSync(listener: (event: ClockSyncEvent) => void): void {
    this.eventListeners.push(listener);
  }

  // -------------------------------------------------------------------------
  // Async Synchronization
  // -------------------------------------------------------------------------

  /**
   * Perform a single async sync. Never blocks the event loop.
   * Uses syncInProgress flag to prevent overlapping syncs.
   */
  private async performAsyncSync(): Promise<void> {
    if (this.syncInProgress) {
      return;
    }

    this.syncInProgress = true;

    try {
      let result: SyncResult;

      if (this.activeProtocol === "ptp") {
        result = this.syncViaPTP();
        if (!result.success) {
          result = await this.asyncSyncViaNTP();
          if (result.success) {
            this.activeProtocol = "ntp";
          }
        }
      } else if (this.activeProtocol === "ntp") {
        result = await this.asyncSyncViaNTP();
      } else {
        result = this.syncViaSystem();
      }

      if (!result.success) {
        result = this.syncViaSystem();
        this.activeProtocol = "system";
      }

      if (result.success) {
        this.applyCorrection(result.offsetMs);
        this.lastSyncAt = Date.now();
        this.lastSyncSuccess = true;
        this.syncCount++;

        // Emit AEP_CLOCK_SYNC event
        const event = createClockSyncEvent({
          bridgeTimeMs: this.now(),
          source: this.activeProtocol,
          offsetMs: result.offsetMs,
          syncedAt: this.lastSyncAt,
        });

        for (const listener of this.eventListeners) {
          try {
            listener(event);
          } catch {
            // Listener errors must not crash sync
          }
        }
      } else {
        this.lastSyncSuccess = false;
        if (typeof console !== "undefined") {
          console.warn(
            "[AsyncBridgeClock] NTP sync failed, retrying on next interval. " +
            "Keeping previous offset:",
            this.currentNtpOffset,
          );
        }
      }
    } catch {
      this.lastSyncSuccess = false;
    } finally {
      this.syncInProgress = false;
    }
  }

  /**
   * Apply a clock correction via slewing or stepping.
   *
   * - |correction| < 1ms: ignore (within noise)
   * - 1ms <= |correction| <= 1000ms: slew over the next sync interval
   * - |correction| > 1000ms: step immediately (log warning)
   *
   * Slewing ensures monotonicity: negative corrections slow the clock
   * rather than moving it backward.
   */
  private applyCorrection(newOffsetMs: number): void {
    const correctionMs = newOffsetMs - this.currentNtpOffset;
    const absCorrectionMs = Math.abs(correctionMs);

    if (absCorrectionMs < SLEW_IGNORE_THRESHOLD_MS) {
      // Within noise: ignore
      return;
    }

    if (absCorrectionMs > SLEW_STEP_THRESHOLD_MS) {
      // Large correction: step immediately
      this.currentNtpOffset = newOffsetMs;
      this.slewRemainingMs = 0;
      this.slewTotalMs = 0;
      if (typeof console !== "undefined") {
        console.warn(
          `[AsyncBridgeClock] Large clock correction applied (step): ${correctionMs.toFixed(1)}ms`,
        );
      }
      return;
    }

    // Slew: spread the correction over the next sync interval
    // Complete any previous slew first
    if (this.slewRemainingMs !== 0) {
      this.currentNtpOffset += this.slewTotalMs;
      this.slewRemainingMs = 0;
    }

    this.slewTotalMs = correctionMs;
    this.slewRemainingMs = correctionMs;
    this.slewStartPerf = performance.now();
    this.slewDurationMs = this.config.syncIntervalMs;
  }

  // -------------------------------------------------------------------------
  // NTP (async, non-blocking)
  // -------------------------------------------------------------------------

  /**
   * SNTP sync via async UDP. Wrapped in a Promise, never blocks event loop.
   */
  private asyncSyncViaNTP(): Promise<SyncResult> {
    const server = this.config.source || DEFAULT_NTP_SERVER;
    const t0 = Date.now();

    return new Promise<SyncResult>((resolve) => {
      let resolved = false;
      const socket = dgram.createSocket("udp4");

      const timeoutHandle = setTimeout(() => {
        if (!resolved) {
          resolved = true;
          socket.close();
          resolve({ success: false, offsetMs: 0, latencyMs: 0 });
        }
      }, NTP_TIMEOUT_MS);

      socket.on("error", () => {
        if (!resolved) {
          resolved = true;
          clearTimeout(timeoutHandle);
          socket.close();
          resolve({ success: false, offsetMs: 0, latencyMs: 0 });
        }
      });

      socket.on("message", (msg: Buffer) => {
        if (resolved) return;
        resolved = true;
        clearTimeout(timeoutHandle);
        const t3 = Date.now();

        if (msg.length < NTP_PACKET_SIZE) {
          socket.close();
          resolve({ success: false, offsetMs: 0, latencyMs: 0 });
          return;
        }

        const t1 = this.extractNTPTimestamp(msg, 32);
        const t2 = this.extractNTPTimestamp(msg, 40);
        const offsetMs = ((t1 - t0) + (t2 - t3)) / 2;
        const latencyMs = t3 - t0;

        socket.close();
        resolve({ success: true, offsetMs, latencyMs });
      });

      const packet = Buffer.alloc(NTP_PACKET_SIZE);
      packet[0] = NTP_REQUEST_HEADER;

      socket.send(packet, 0, NTP_PACKET_SIZE, NTP_PORT, server, (err) => {
        if (err && !resolved) {
          resolved = true;
          clearTimeout(timeoutHandle);
          socket.close();
          resolve({ success: false, offsetMs: 0, latencyMs: 0 });
        }
      });
    });
  }

  private extractNTPTimestamp(buffer: Buffer, byteOffset: number): number {
    const seconds = buffer.readUInt32BE(byteOffset);
    const fraction = buffer.readUInt32BE(byteOffset + 4);
    const unixSeconds = seconds - NTP_UNIX_EPOCH_DELTA_SECONDS;
    const fractionalMs = (fraction / 0x100000000) * 1000;
    return (unixSeconds * 1000) + fractionalMs;
  }

  // -------------------------------------------------------------------------
  // PTP (synchronous, fast)
  // -------------------------------------------------------------------------

  private syncViaPTP(): SyncResult {
    try {
      const raw = fs.readFileSync(PTP_OFFSET_PATH, "utf-8").trim();
      const nanoseconds = parseInt(raw, 10);
      if (isNaN(nanoseconds)) {
        return { success: false, offsetMs: 0, latencyMs: 0 };
      }
      return { success: true, offsetMs: nanoseconds / 1_000_000, latencyMs: 0 };
    } catch {
      return { success: false, offsetMs: 0, latencyMs: 0 };
    }
  }

  // -------------------------------------------------------------------------
  // System fallback
  // -------------------------------------------------------------------------

  private syncViaSystem(): SyncResult {
    if (typeof console !== "undefined") {
      console.warn(
        "[AsyncBridgeClock] Using system clock fallback - no external time source available.",
      );
    }
    return { success: true, offsetMs: 0, latencyMs: 0 };
  }
}
